finesse/hardware/data_file_writer.py

```diff
@@ -0,0 +1,117 @@
+"""Provides a class for writing sensor data to a CSV file."""
+import platform
+from datetime import datetime
+from decimal import Decimal
+from math import floor
+from pathlib import Path
+from typing import Any
+
+from csvy import Writer
+from pubsub import pub
+
+from .. import config
+from .pubsub_decorators import pubsub_errors
+from .stepper_motor import get_stepper_motor_instance
+from .temperature import get_hot_bb_temperature_controller_instance
+
+
+def _on_error_occurred(error: BaseException) -> None:
+    """Close file in case of error."""
+    pub.sendMessage("data_file.close")
+
+
+def _get_platform_info() -> dict[str, str]:
+    return {
+        key: getattr(platform, key)() for key in ("platform", "python_version", "node")
+    }
+
+
+def _get_metadata(filename: str) -> dict[str, Any]:
+    return {
+        "encoding": "utf-8",
+        "name": filename,
+        "datetime": datetime.now().astimezone().isoformat(),  # include timezone
+        "system": {
+            "app": {
+                "name": config.APP_NAME,
+                "author": config.APP_AUTHOR,
+                "version": config.APP_VERSION,
+            },
+            "platform": _get_platform_info(),
+        },
+    }
+
+
+class DataFileWriter:
+    """A class for writing sensor data to a CSV file.
+
+    This class is a singleton. Every time a new file needs to be written this instance
+    is reused.
+    """
+
+    def __init__(self) -> None:
+        """Create a new DataFileWriter."""
+        self._writer: Writer
+        """The CSV writer."""
+
+        # Listen to open/close messages
+        pub.subscribe(self.open, "data_file.open")
+        pub.subscribe(self.close, "data_file.close")
+
+        # Listen for error messages
+        pub.subscribe(_on_error_occurred, "data_file.error")
+
+    @pubsub_errors("data_file.error")
+    def open(self, path: Path) -> None:
+        """Open a file at the specified path for writing.
+
+        Args:
+            path: The path of the file to write to
+        """
+        self._writer = Writer(path, _get_metadata(path.name))
+
+        # Write column headers
+        self._writer.writerow(
+            (
+                "Date",
+                "Time",
+                *(f"Temp{i+1}" for i in range(config.NUM_TEMPERATURE_MONITOR_CHANNELS)),
```

Comment:
Do we want to add a format specifier here? Otherwise it will write the `Decimal`s with silly numbers of decimal places